src/RT_RRT_star.cpp

```diff
@@ -54,25 +54,227 @@ namespace rrt {
+    double c_min = cost(closest) + dist(closest, rand);
     for (typename std::vector<Utils::Point<T> >::iterator itr = x_near.begin();
       itr != x_near.end(); ++itr) {
 
-      int c_new = cost(*itr) + dist(*itr, rand);
+      double c_new = cost(*itr) + dist(*itr, rand);
+
+    }
+
+  }
+
+  template <class T>
+  double dist(Utils::Point<T> first, Utils::Point<T> second)
+  {
+    return sqrt(pow(first.x-second.x,2)+pow(first.y-second.y,2));
+  }
+
+  template <class T>
+  void cube_round(float xyz[], float rxyz[])
+  {
+    float rx = nearbyint(xyz[0]);
+    float ry = nearbyint(xyz[1]);
+    float rz = nearbyint(xyz[2]);
+
+    float x_diff = fabs(rx - xyz[0]);
+    float y_diff = fabs(ry - xyz[1]);
+    float z_diff = fabs(rz - xyz[2]);
+
+    if ((x_diff > y_diff) && (x_diff > z_diff))
+        rx = -ry-rz;
+    else if (y_diff > z_diff)
+        ry = -rx-rz;
+    else
+        rz = -rx-ry;
+    rxyz[0] = rx; rxyz[1] = ry; rxyz[2] = rz;
+  }
+
+  template <class T>
+  std::pair<int, int > cube_to_axial(float rxyz[])
+  {
+      int q = rxyz[0];
+      int r = rxyz[2];
+      return std::pair<int, int >(q, r);
+  }
+
+  template <class T>
+  void axial_to_cube(float q, float r, float xyz[])
+  {
+    xyz[0] = q;
+    xyz[2] = r;
+    xyz[1] = -xyz[0]-xyz[2];
+  }
+
+  template <class T>
+  std::pair<int, int > hex_round(float q,float r)
+  {
+    float xyz[3];
+    RT_RRT<T>::axial_to_cube(q,r,xyz);
+    float rxyz[3];
+    RT_RRT<T>::cube_round(xyz, rxyz);
+    return RT_RRT<T>::cube_to_axial(rxyz);
+  }
+
+  template <class T>
+  std::pair<int, int> Grid_Id(Utils::Point<T> gride_idx, int size = 4)
+  {
+    //size is edge length of hexagon
+    float x = gride_idx.x, y = gride_idx.y;
+    float q = (x * sqrt(3)/3 - y / 3) / size;
+    float r = (y * 2/3) / size;
+    std::pair<int, int> here = RT_RRT<T>::hex_round(q,r);
+    return here;
+  }
+
+  template <class T>
+  std::pair<double,Utils::Point<T> > cost(Utils::Point<T> child, int count=0, int k=-1)
+  {
+            // k is location in the tree
+            if (child==RT_RRT<T>::Xa)
+              return std::pair<double,Utils::Point<T> > (0,RT_RRT<T>::Xa);
+            std::pair<double,Utils::Point<T> > here;
+            if (count==0)
+            {
+              int j=0;
+              for(;j<RT_RRT<T>::tree.size();j++)
+              {
+                if(RT_RRT<T>::tree[j].first==child)
+                  break;
+              }
+              if (RT_RRT<T>::line_path_obs(child,(RT_RRT<T>::tree[RT_RRT<T>::tree[j].second].first)))
+                here = cost(RT_RRT<T>::tree[RT_RRT<T>::tree[j].second].first,1,RT_RRT<T>::tree[j].second);
+              else
+                return std::pair<double,Utils::Point<T> > (std::numeric_limits<double>::infinity(),RT_RRT<T>::tree[RT_RRT<T>::tree[j].second].first);
+            }
+            else
+            {
+              if (RT_RRT<T>::line_path_obs(child,(RT_RRT<T>::tree[RT_RRT<T>::tree[j].second].first)))
```

Comment:
This seems to cause build error, did you mean to put this within the `for loop` ?

Reply:
Oops! It should have been `k`, not `j`. :stuck_out_tongue: